### PR TITLE
IOperation Refactor Pt 1

### PR DESCRIPTION
**Pull request by 333fred**
Start of the IOperation changes. I copied the existing writer, made some small changes to support running both at the same time, and did a small bit of refactoring to move to .NET 5 and fully nullable enable. I then introduced a base class between Operation and the existing node types, and pulled the functionality I will be removing from Operation into that new class.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a quick review. I'd recommend commmit-by-commit for easiest viewing.

**Comment by 333fred** [comment]
Actually, hold off on reviewing until I do some commit reordering and squashing tomorrow.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler this should be ready for review now. Again, I recommend commit-by-commit here.

File: src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs
Comment (sharwell) [inline src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:138]: ❔ In what cases is this null?

Comment (333fred) [reply src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:138]: When you're operating on operations that come from a CFG. In the original nodes, it's never null.

Comment (333fred) [reply src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:138]: Beyond docs (which we have on the property) I don't know of a good way to annotate that other than `?` and suppress/assert the warnings away.

File: src/Compilers/Core/Portable/Operations/IOperation.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/Operations/IOperation.cs:69]: 😦 This may be rough for usability

File: src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchStatementCodeFixProvider.cs
Comment (sharwell) [inline src/Analyzers/Core/CodeFixes/PopulateSwitch/AbstractPopulateSwitchStatementCodeFixProvider.cs:64]: 💭 Maybe better to do something like this?

```suggestion
            => switchOperation.Value.Type ?? throw ExceptionUtilities.Unreachable;
```

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for reviews please.

File: Compilers.sln
Comment (jasonmalinowski) [inline Compilers.sln:67]: What's happening to the old one?

Comment (333fred) [reply Compilers.sln:67]: I'm porting type by type. When I'm done, I'll kill the old one and rename this back.

**Comment by genlu** [comment]
Just curious, what's the ultimate goal of this?

**Comment by 333fred** [comment]
> Just curious, what's the ultimate goal of this?

@genlu added you to the email thread, but the ultimate goal will be to remove laziness from IOperation. It does not buy us any meaningful perf benefits, while complicating the implementation and introducing race conditions that are very hard to fix. I'll be going type-by-type to port them to a non-lazy, shared implementation with some structural improvements, then refactoring how the semantic model gets the tree in order to make the whole process much better.

File: src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs:11]: It would be preferable to update the generated file in place, rather than move it chunk by chunk. That would allow reviewers to see the diff.

Comment (333fred) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs:11]: I don't think that will be easier for review, personally.

Comment (jcouv) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs:11]: For starters, it would make it clearly visible whether the OperationVisitor changed at all. As the PR stands, there is no easy way to tell.

---
In reply to: [501307338](https://github.com/dotnet/roslyn/pull/48226#discussion_r501307338) [](ancestors = 501307338)

Comment (333fred) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs:11]: OperationVisitor won't be changing. It's a copy-paste currently, and it won't be modified until the very end when I nullable enable everything else.

As to other types, the reason I don't think it will be easier to review is that, particularly when we get to the hierarchies of Base/NonLazy/Lazy types, the diff will end up being far too cluttered to read, as git tries to relate new lines to old lines. They're not really related, and for the most part are uninteresting. The interesting bits are in the files that are not generated.

Comment (jcouv) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs:11]: If the removal of some types (Lazy or Base) cause diff to get confused, then the PR should be split into two commits (one that just removes types, one that makes the updates to remaining types).
Starting a new file guarantees that we can't diff, which is definitely worse.

> The interesting bits are in the files that are not generated.

Do you have an example of the change you plan to for the remaining generated code? That might help inform the discussion.

---
In reply to: [501311258](https://github.com/dotnet/roslyn/pull/48226#discussion_r501311258) [](ancestors = 501311258)

Comment (333fred) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs:11]: > If the removal of some types (Lazy or Base) cause diff to get confused, then the PR should be split into two commits (one that just removes types, one that makes the updates to remaining types).

This is not the approach I'm taking. The approach I'm taking is that every commit will change 1 type. This makes the diff very reviewable. I just don't think the diff in this file is particularly useful, and will necessitate adding a bunch of logic to the operation writer around nullable enabling and then disabling for some contexts.

> Do you have an example of the change you plan to for the remaining generated code?

Sure, here's one example: https://github.com/dotnet/roslyn/commit/845e602db46b999e3f1eb18b912c1b45636367bb. Every commit will look something like this.

Comment (jcouv) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs:11]: Thanks for the example commit. The diff in that example would be small and easy to review with update-in-place. 
In that example, the interface type is unchanged except maybe for `#nullable` around it and two properties added to actual type. The diff algorithm will keep up fine (such changes are what most of our PRs do). 

If you want to try a diff on a more complex change (with removal of Lazy type), that's an easy experiment. 
I even think we need to include at least a type or two in this first PR to exercise all the code you're adding.

The reason I'm insisting is because I know that many more PRs/commits/types are following. A small investment on the first round will make many reviews easier.

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs
Comment (jaredpar) [inline src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs:75]: What test above is giving us confidence that `Instance.Type` is non-null here?

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs
Comment (jaredpar) [inline src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs:79]: Do u anticipate lots `SemanticModel!.` uses in this series of changes? If so should we consider a model like the IDE has where we introduce say `RequiredSemanticModel`?

File: src/Compilers/Core/Portable/Operations/OperationOld.cs
Comment (jaredpar) [inline src/Compilers/Core/Portable/Operations/OperationOld.cs:9]: Consider adding a `PROTOTYPE` comment to make sure we remember to delete this before merging to master

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGeneratorNew/IOperationClassWriter.PortedTypes.cs
Comment (jaredpar) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGeneratorNew/IOperationClassWriter.PortedTypes.cs:11]: Use an explicit string comparer here.

Comment (333fred) [reply src/Tools/Source/CompilerGeneratorTools/Source/IOperationGeneratorNew/IOperationClassWriter.PortedTypes.cs:11]: Does this matter? It's not sticking around.

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs
Comment (333fred) [reply src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs:79]: I anticipate no more than what was required to get this to compile.

File: src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs
Comment (alrz) [reply src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:138]: The only valid use of SemanticModel is when it does NOT come from CFG, that is, it's still a bug if you check for null in that case, because it's always null. I think we should return `SemanticModel` and throw if it's not available.

File: src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs:11]: >  I even think we need to include at least a type or two in this first PR to exercise all the code you're adding.

I'm not going to put type ports in this PR. Yes, there are a couple of bugs in the generator as written, but they're already fixed and were fixed inline with the commit that discovered them.

Comment (jcouv) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs:11]: Since when do we merge code that is not exercised? Is there a reason this PR is special?

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs
Comment (333fred) [reply src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs:75]: The fact that `null` can never be an instance for an invocation, and therefore the type will never be `null`.

File: src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs
Comment (333fred) [reply src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:138]: > I think we should return SemanticModel and throw if it's not available.

This is a behavior change, and a silently breaking one that there is no way to work around. I'm not interested in making actual behavior changes as part of this series of PRs. The fact is that it can return `null`, and that's what it needs to be annotated as.

File: src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs:11]: > Since when do we merge code that is not exercised? Is there a reason this PR is special?

There's nothing new in this PR, no new code. The code that is here is tested, or will be rewritten entirely.

File: src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs
Comment (mavasani) [reply src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:138]: Is there a known use case where this would be a _real_ breaking change? I am pretty sure this field is not accessed anywhere in the roslyn-analyzers repo when the operation is obtained from the CFG. I agree that adding these null-forgiving operators as all use sites is going to be extremely confusing and burdensome.

Comment (alrz) [reply src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:138]: >This is a behavior change, and a silently breaking one that there is no way to work around. I'm not interested in making actual behavior changes as part of this series of PRs. The fact is that it can return null, and that's what it needs to be annotated as.

@333fred Then we'll have to use forgiving operator on ALL actual valid usages. While we don't even use this property when it's always null.

File: src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs
Comment (jcouv) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs]: > There's nothing new in this PR, no new code.

You're losing me entirely. The old generator has added logic for `PortedTypes` and the new generated code produced different output in the example you shared (except that it's not exercised in this PR, which is a problem).
This discussion is all the more reason to just update the generator and generated files in place. Then it will be clear what is changing. 
I'm expecting the generator to add logic for `if (typeIsPorted) { ... generate new code ... } else { ... generate old code ... }`. The rest of the generator I assume is unchanged, but could not verify easily because of duplication.

File: src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs
Comment (333fred) [reply src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:138]: If we want to consider a breaking API change, we can do that in the future. I'm specifically not making any intentional behavior changes as part of the series of PRs here.

File: src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Generated/Operations.new.Generated.cs]: > new generated code produced different output in the example you shared (except that it's not exercised in this PR, which is a problem).

It did not. There are no logic changes in this PR beyond copy/paste and including/excluding types. The actual changes to the format are going to be in the next PR. Regardless, I've spent most of the day redoing the generator logic and subsequent commits to do in-place transformation, as opposed to side-by-side transformation. I don't think it's particularly nicer or more reviewable, but it'll be what you want.

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/Program.cs
Comment (jcouv) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/Program.cs:32]: >Derserialize [](start = 23, length = 12)

nit: typo

**Review by jcouv — APPROVED** [review]
LGTM. Thanks much for the effort :-)

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/Program.cs
Comment (333fred) [reply src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/Program.cs:32]: No no, it's German. I swear...

**Comment by 333fred** [comment]
@jaredpar for another review please.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler can I please have a second review

File: src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs
Comment (jaredpar) [reply src/Analyzers/Core/Analyzers/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs:138]: Agree with @333fred here. We need to be careful about separating nullable annotation passes (essentially documenting the existing behavior of our APIs) with changes that break those behaviors to better fit user scenarios. Small changes sure we make during the nullable annotation passes. But customer visible impactful changes should be considered as a separate change entirely so we can evaluate it on its individual merits

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs
Comment (jaredpar) [reply src/Analyzers/Core/Analyzers/SimplifyInterpolation/Helpers.cs:75]: Hmm, wish we had a way of annotating such a relationship here but we don't :(

**Review by jaredpar — APPROVED** [review]

**Comment by 333fred** [comment]
@sharwell or @jasonmalinowski, any further comments?

**Comment by jasonmalinowski** [comment]
@333fred I had just looked out of curiosity earlier and left the random comment. Don't consider me an actual reviewer, more the peanut gallery.

**Comment by 333fred** [comment]
@jasonmalinowski well I do need an IDE sign off :wink:

**Review by jasonmalinowski — APPROVED** [review]

**Comment by jasonmalinowski** [comment]
@333fred Oh in that case, I can kinda pretend I know what I'm doing.

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs
Comment (alrz) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs:161]: @333fred So if this is being removed, where it's generated?

Comment (333fred) [reply src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs:161]: https://github.com/dotnet/roslyn/blob/main/src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs#L283.

Comment (alrz) [reply src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs:161]: Yes, but that's not called anywhere. Is this intentional?

Comment (333fred) [reply src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs:161]: Uh, no, it's not. Thanks for catching this, I'll go fix.

Comment (333fred) [reply src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs:161]: https://github.com/dotnet/roslyn/pull/53812/

**Pull request merged**
